Feature/import template

Jacarte opened the pull request:
The class **ImportGenerator** generates the text template as a scala trait. It contains two main methods: **generateImportText** and **createScalaProjectForImports**.
- **generateImportText**: Given a vector of *Import* instances, creates the text template as scala trait. The key insight is to write this text to a file and then implement it.  
   ```
    val text = generator.generateImportText(modul1.imports.concat(modul2.imports))
   ```

- Creates/Modify a module boilerplate. The created module contains a scala file with the content obtained from the mentioned method **generateImportText**.  
  ```
   generator.createScalaProjectForImports("C_WRAPPER", modul1.imports)
   ```
  The module structure is following the example above is: ```C_WRAPPER/src/GeneratedImports.scala```

[See](https://github.com/Jacarte/swam/blob/feature/import_template_engine1/runtime/test/src/generator/TemplateGeneratorTests.scala) for a full example.

satabin commented:
Hi @Jacarte 

Thanks, it looks nice. However, I am not a big fan f having code generation in a library. I think this would be better in a separate module than in the `runtime` one. Maybe a new top-level module named `generator` which might even have a main class to be run from command line.

Generally speaking, plain old code generation in text file is not really idiomatic in Scala. For the best as for the worst, Scala tends to do that with macros. The good part of macros is that you get nice compiler reporting in case of type error, and you can relate them to your source code directly. The bad part of macros is that they are macros. I won't ask you to write the macros (I am not that sadistic) but I might add it to cover the same use case as you do with a different approach. Your approach is fine in a typical source generation in a build tool (be it maven, gradle, sbt, mill, or whatever).

I will review in more details, but please first consider extracting the code to a specific module.

satabin commented on runtime/src/swam/runtime/import_generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,182 @@
+package swam.runtime.import_generator
```
For a reason I cannot really explain, I really dislike underscores in package names, as well as camel case. I like my packages to be one word, or if it is really necessary several words, all lower case. In this case I think `swam.runtime.generator` is an ok name.

satabin commented on runtime/src/swam/runtime/import_generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,182 @@
+      if (t.length == 1)
+        getScalaType(t.last)
+      else
+        s"(${t.map(getScalaType).mkString(",")})"
```
```suggestion
        t.map(getScalaType(_).mkString("(", ",", ")")
```

satabin commented on runtime/src/swam/runtime/import_generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,182 @@
+/**
+  * Generates the scala code with the template to use as Import in the WASM execution
+  */
+class ImportGenerator() {
```
Parentheses are useless here

satabin commented on runtime/src/swam/runtime/import_generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,182 @@
+  */
+/**
+  * Generates the scala code with the template to use as Import in the WASM execution
+  */
```
I guess the double scaladoc comes from the IDE. Can you please merge them?

satabin commented on runtime/src/swam/runtime/import_generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,182 @@
+      case I64 => s"$index:${getScalaType(tpr)}"
+      case F32 => s"$index:${getScalaType(tpr)}"
+      case F64 => s"$index:${getScalaType(tpr)}"
+    }}"
```
Why do you need all these cases? I think this does the same:
```suggestion
    s"p$index:${getScalaType(tpr)}"
```

Jacarte commented:
Hi @satabin 

Thanks for your quick reply :). I will move it to a different module giving a try to the macros, it seems a nice learning exercise. I will let you know.

Cheers

satabin commented on runtime/src/swam/runtime/import_generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,182 @@
+      .zip(Seq.range(0, func.params.length))
+      .map(t => {
+        buildParameter(t._2, t._1)
+      })
```
A better way of doing this is probably along these lines:
```suggestion
    s"def $name(${func.params
      .zipWithIndex
      .map {
        case (p, idx) => buildParameter(idx, p)
      }
```

satabin commented on runtime/src/swam/runtime/import_generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,182 @@
+    * @param imports
+    * @return
+    */
+  def generateImportText(imports: Vector[Import]) = {
```
This looks awfully imperative to me ^^ Using a `StringBuilder` instead of `var` and string concatenation would be already better.

Jacarte replied:
What do you think about a template engine? I really prefer to use one in these cases.

satabin commented on runtime/test/src/generator/TemplateGeneratorTests.scala:
```diff
@@ -0,0 +1,73 @@
+/**
+  * @author Javier Cabrera-Arteaga on 2020-03-06
+  */
+object TemplateGeneratorTests extends TestSuite {
```
These tests don't really look like tests to me, I would move this into example code instead.

satabin commented on runtime/test/trace/src/TracerTests.scala:
```diff
@@ -0,0 +1,51 @@
+
+import utest._
+
+object TracerTests extends TestSuite {
```
Is that a leftover from another change? I thought we already had this kind of tests.

Jacarte replied:
It seems so ... to remove then :)

Jacarte commented:
Hi @satabin 

Thanks for the tips, all useful :)

satabin commented:
Hi @Jacarte 
I received a notification where you ask about template engines, but I can't find your comment in the PR. So,  I am totally in favor of it. I personally like Mustache, as it is simple and does templating well without complex structures. This would be a good use case for this, so go for it if you feel like it. As a nice side effect, it makes it easier for potential users to customize generated code.

Jacarte commented:
> Hi @Jacarte
> I received a notification where you ask about template engines, but I can't find your comment in the PR. So, I am totally in favor of it. I personally like Mustache, as it is simple and does templating well without complex structures. This would be a good use case for this, so go for it if you feel like it. As a nice side effect, it makes it easier for potential users to customize generated code.

Hi @satabin 

I moved the implementation to a new module (generator). Added some packages, in this case, *sopt* for command line parsing, *scalate* for the template engine and *scalafmt* to check if the generated file is a valid scala file.

The default template is in the resource folder of the module (as a mustache file). Also, in the command line you can specify if you want to output the json map to be used with other purposes.

I plan to add some other features. I'll let you know.

Thanks in advance (I think this version is a little bit less imperative 👍 )

satabin commented on generator/src/swam/generator/Generator.scala:
```diff
@@ -0,0 +1,90 @@
+/**
+    @author Javier Cabrera-Arteaga on 2020-03-07
+  */
+object Generator extends App {
```
You should use [`IOApp`](https://typelevel.org/cats-effect/datatypes/ioapp.html), which is intended exactly to have an `IO` based main, with all the `ContextShift` and friends you want to have.

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,150 @@
+    val sorted = imports
+      .collect { case i: Import.Function => i } // Filter by function type
+      .groupBy(t => t.moduleName) // Group by module
+      .map[String, Set[Import.Function]](t => (t._1, t._2.toSet)) // remove duplicated entries
```
```suggestion
      .view
      .mapValues(_.toSet) // remove duplicated entries
```

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,150 @@
+                  "imports" -> getContext(imports)
+                ))
+
+    val file = Paths.get(s"GeneratedImport.scala")
```
Don't you want to make the file name a CLI parameter?

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,150 @@
+
+    val file = Paths.get(s"GeneratedImport.scala")
+
+    try {
```
It would be more idiomatic with the project to suspend side-effects in an `IO` and to not use `try`/`catch` but the `IO` error handling functions.

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,150 @@
+    */
+  def createScalaProjectForImports(projectName: String, imports: Vector[Import]) = {
+    s"mkdir -p $projectName".!!
+    s"mkdir -p $projectName/src".!!
```
There are better ways of doing this, in our case we have [`fs2-io`](https://oss.sonatype.org/service/local/repositories/releases/archive/co/fs2/fs2-io_2.12/2.2.1/fs2-io_2.12-2.2.1-javadoc.jar/!/fs2/io/file/index.html)

Jacarte replied:
Solved

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,150 @@
+
+    val printer = new PrintWriter(new File(s"$projectName/src/GeneratedImport.scala"))
+    printer.print(trait_)
+    printer.close()
```
Again, using fs2-io here to write in the file would be more reliable.

Jacarte replied:
Moved to fs2io

satabin commented on generator/test/src/generator/TemplateGeneratorTests.scala:
```diff
@@ -0,0 +1,62 @@
+/**
+  * @author Javier Cabrera-Arteaga on 2020-03-06
+  */
+object TemplateGeneratorTests extends TestSuite {
```
Still not convinced this is a test. There is no check performed in the test cases.

Jacarte replied:
Removed test

satabin commented on generator/src/swam/generator/Generator.scala:
```diff
@@ -0,0 +1,90 @@
+  implicit val cs = IO.contextShift(ExecutionContext.Implicits.global)
+
+  val blockingPool = ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())
+  val blocker: Blocker = Blocker.liftExecutionContext(blockingPool)
```
You should use the [`Resource`](https://typelevel.org/cats-effect/datatypes/resource.html) based `Blocker` as demonstrated [here](https://typelevel.org/cats-effect/datatypes/contextshift.html#blocker).

Jacarte replied:
I think is fixed. But, honestly, I think I am not understanding cats very wel

satabin replied:
The idea of cats-effect is that you can write code with side-effects in a pure way. Since this is intended to be integrated into swam, I'd like to have the code up to the standards of the rest of the code. If you don't feel like you get it enough, I can take your changes and rewrite the parts that need to be improved.

Jacarte commented:
Hi @satabin 

I think all is fixed now

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,163 @@
+                  )
+                }
+              )
+          ))
```
When dealing with tuples, it is a bit more elegant (I think) to use the partial function syntax to deconstruct the parameters. Something like this (not compiled): 
```suggestion
      .map {
        case (name, idx) =>
          Map(
            "module" -> name,
            "comma" -> (idx < sorted.keys.size - 1),
            "fields" -> sorted(name).toSeq.zipWithIndex
              .map {
               case (Function(moduleName, fieldName, tpe), idx) =>
                  Map(
                    "name" -> fieldName,
                    "return" -> buildReturn(tpe.t),
                    "params" -> tpe.params.zipWithIndex.map(t => s"p${t._2}:${getScalaType(t._1)}").mkString(","),
                    "comma" -> (idx < sorted(name).size - 1)
                  )
              }
          )
        }
```

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,163 @@
+      .mapValues(_.toSet) // remove duplicated entries
+
+    // Generating DTO
+    val dto = sorted.keys.zipWithIndex
```
Since you use the key to access element in `sorted`, I think here you can simply use `zipWithIndex` on the map to get the key and value at once:
```scala
sorted.zipWithIndex.map {
  case ((name, functions), idx) =>
    ...
}
```

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,163 @@
+
+    Blocker[F]
+      .use { blocker =>
+        {
```
Useless inner braces.

satabin commented on generator/src/swam/generator/Generator.scala:
```diff
@@ -0,0 +1,112 @@
+    if (config.createBoilerplate.isEmpty) {
+      println()
+      if (!config.printTemplateContext)
+        IO(println(generator.generateImportText(imports, config.className)))
```
I love seeing all these side-effects wrapped in an `IO`! Ideally the one above would be as well, and then chained with this one.

satabin commented on generator/src/swam/generator/Generator.scala:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  def run(args: List[String]): IO[ExitCode] = parser.parse(args, Config()) match {
+    case Some(config) => {
```
The braces in the cases are useless. I prefer to put parentheses around the `for` comprehension in this case.

In scala, braces and parentheses really have different semantics, and over-using braces to wrap simple expressions can lead to hard to detect bug, especially when using suspended effects.

Consider this snippet:

```scala
val io: IO[Unit] = {
  IO(println("print1"))
  IO(println("print2"))
}
```

when running `io`, for instance with `unsafeRunSync()`, the first IO is never executed. In the block defining `io` you have 2 imperative statements, and the first one just creates an `IO` that will never gets executed.

If you need to sequence `IO`s while disregarding previous result (the case in your `println`s), use the `>>` operator:

```scala
val io: IO[Unit] =
  IO(println("print1")) >> IO(println("print2"))
```

Now when you run `io` you see the two outputs in the correct order. If you need to sequence while using previous result, then you already know it, it's the `for` comprehension as below.

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -0,0 +1,163 @@
+  val config = Paths.get(".scalafmt.conf")
+
+  private var templateFile = getClass.getClassLoader.getResource("import_template.mustache").getFile
+  var registeredFunctions: Set[String] = Set[String]()
```
It would be great if there were no `var` at all. The `registeredFunctions` one seems not to be used anymore anyway.

satabin commented:
Nice work! I am really happy to see that you are getting how to use the `IO` in this project. Still have a few comments to get to perfection :)

Jacarte commented:
> Nice work! I am really happy to see that you are getting how to use the `IO` in this project. Still have a few comments to get to perfection :)

Thanks @satabin ! glad to hear it, I think I am getting it. But the thing is I have a good mentor 🥇

satabin commented on generator/src/swam/generator/ImportGenerator.scala:
```diff
@@ -23,28 +23,20 @@ class ImportGenerator[F[_]: Effect](implicit cs: ContextShift[F]) {
-
-  private var templateFile = getClass.getClassLoader.getResource("import_template.mustache").getFile
-  var registeredFunctions: Set[String] = Set[String]()
+  val defaultTemplate = getClass.getClassLoader.getResource("import_template.mustache").getFile;
```
Almost ^^ can you remove the semicolon?

satabin commented:
Looks good, let's merge it. Thanks for the contribution. I will publish a new snapshot with it immediately

satabin commented:
https://oss.sonatype.org/content/repositories/snapshots/org/gnieh/swam-generator_2.13/

Jacarte commented:
Nice! I am preparing for another PR. Hopefully, I learned something here:) and this will be better.

satabin commented:
Sure, no worries, if you are in no rush, we have all the time to discuss on the PR :)

Jacarte commented:
BTW, I will write an example to be published on the web page.
